File: src/56. Merge Intervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        List<Interval> res = new ArrayList<>(); 
        int n = intervals.size();
        int[] a = new int[n];
        int[] b = new int[n];
        for (int i = 0; i < n; i++){
            a[i] = intervals.get(i).start;
            b[i] = intervals.get(i).end;
        }
        Arrays.sort(a);
	    Arrays.sort(b);
        for (int i = 0, j = 0; i < n; i++) {
            // j is start of interval, i is the end
            // 如果i指向a数组中的最后一个位置，或者a数组i+1位置上的数字大于b数组i位置的数时，
            // 此时说明区间已经不连续了
		    if (i == n - 1 || a[i + 1] > b[i]) {
			    res.add(new Interval(a[j], b[i]));
			    j = i + 1;
		    }
	    }
	    return res;
    }
}

File: cpp/56. Merge Intervals.cpp
// 如1-6 2-4 3-8 -> 1-4 2-6 3-8 merge后的区间与把开始阶段，结束阶段的时间都sort，merge结果是不变的   -> 某种可证明的特性
// 如果start的i+1大于end的i，则说明前面的区间是独立的，和后面不重叠

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> res;
        vector<int> s;
        vector<int> e;
        for (int i = 0; i < intervals.size(); i++) {
            s.push_back(intervals[i][0]);
            e.push_back(intervals[i][1]);
        }
        sort(s.begin(), s.end());
        sort(e.begin(), e.end());
        for (int i = 0, j = 0; i < intervals.size(); i++) {
            if (i == intervals.size() - 1 || s[i + 1] > e[i]) {
                res.push_back({s[j], e[i]});
                j = i + 1;
            }
        }
        return res;
    }
};
